zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -504,60 +536,56 @@ impl<F: Field> SignVerifyChip<F> {
+            let mut terms_hi = pk_hash[N_BYTES_WORD - N_BYTES_ACCOUNT_ADDRESS..16]
+                .iter()
+                .rev()
+                .zip(powers_of_256.clone().into_iter())
+                .map(|(byte, coeff)| {
+                    maingate::Term::Unassigned(Value::known(F::from(*byte as u64)), coeff)
+                })
                 .collect_vec();
-            let terms = pk_hash[12..]
+            terms_hi.reverse();
+
+            let mut terms_lo = pk_hash[16..]
                 .iter()
-                .zip(powers_of_256.into_iter().rev())
-                .map(|(byte, coeff)| maingate::Term::Unassigned(*byte, coeff))
+                .rev()
+                .zip(powers_of_256.into_iter())
+                .map(|(byte, coeff)| {
+                    maingate::Term::Unassigned(Value::known(F::from(*byte as u64)), coeff)
+                })
                 .collect_vec();
-            let (address, pk_hash_lo) = main_gate.decompose(ctx, &terms, F::ZERO, |_, _| Ok(()))?;
-
-            (
-                address,
-                pk_hash_lo
-                    .into_iter()
-                    .zip(pk_hash[12..].iter())
-                    .map(|(assigned, byte)| Term::assigned(assigned.cell(), *byte))
-                    .collect_vec(),
-            )
+            terms_lo.reverse();
+
+            // gate
+            let (address_hi, _) = main_gate.decompose(ctx, &terms_hi, F::ZERO, |_, _| Ok(()))?;
+            let (address_lo, _) = main_gate.decompose(ctx, &terms_lo, F::ZERO, |_, _| Ok(()))?;
+
+            Word::new([address_lo, address_hi])
         };
-        let is_address_zero = main_gate.is_zero(ctx, &address)?;
+
+        let iz_zero_hi = main_gate.is_zero(ctx, &address.hi())?;
+        let iz_zero_lo = main_gate.is_zero(ctx, &address.lo())?;
+        let is_address_zero = main_gate.and(ctx, &iz_zero_lo, &iz_zero_hi)?;
 
         // Ref. spec SignVerifyChip 3. Verify that the signed message in the ecdsa_chip
-        // with RLC encoding corresponds to msg_hash_rlc
-        let msg_hash_rlc = {
-            let zero = main_gate.assign_constant(ctx, F::ZERO)?;
-            let assigned_msg_hash_le = assigned_ecdsa
-                .msg_hash_le
-                .iter()
-                .map(|byte| main_gate.select(ctx, &zero, byte, &is_address_zero))
-                .collect::<Result<Vec<_>, _>>()?;
+        // corresponds to msg_hash
+        let msg_hash = {
+            main_gate.assign_constant(ctx, F::ZERO)?;
             let msg_hash_le = (!padding)
                 .then(|| sign_data.msg_hash.to_bytes())
-                .unwrap_or_default()
-                .map(|byte| Value::known(F::from(byte as u64)));
-            self.assign_rlc_le(
-                config,
-                ctx,
-                chips,
-                "msg_hash",
-                config.q_rlc_evm_word,
-                challenges.evm_word(),
-                assigned_msg_hash_le
```

Comment:
Follow @ed255 's hint, it seems copy constrain also lost here for `msg_hash_le`. Means the message_hash can be different with verfified in ecdsa. 

I think the summary pattern here is: for `assign_word`, the input should take `Term<F>` instead of raw bytes, Since the `Term<F>` impliclitly carry copy constrain

Reply:
@ed255 Wondering if you have any solutions for this? Have discussed with sm and have some ideas but don't know how to implement it.

The issue is `assigned_ecdsa.msg_hash_le` has 32 cells but `msg_hash` is a word type which only has 2 cells. Her are some thoughts
1. aggregate 32 cells to 2 cells and do `constrain_equal()` with `msg_hash` 
``` rust
let msg_hash = self.assign_word(config, ctx, msg_hash_le)?;
```
2. aggregate cells through `main_gate.decompose` like, 
```rust
let (msg_hash_lo, ecdsa_new_msg_hash_lo_le) = main_gate.decompose(assigned_ecdsa.msg_hash_le[0:16])
let (msg_hash_hi, ecdsa_new_msg_hash_hi_le) = main_gate.decompose(assigned_ecdsa.msg_hash_le[16:32])
```
3. assign `sign_data.msg_hash` to 32 cells and do `constrain_equal()` with `assigned_ecdsa.msg_hash_le` 
4. get the value of each byte from `msg_hash`, and compare with the value of `assigned_ecdsa.msg_hash_le`
---


For 1, no ideas how to do the cell aggregation and I don't think it's doable bcs there are other info like, `region_index` and `row_offset` inside.
For 2, wil have extra assigments
For 3, also have extra assigments
For 4, constratins are usually for cells or expressions, not for values

Reply:
I think @hero78119 is right here, well spotted!

I think the best option is (2). You can follow this pattern from the commit I shared in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1418#discussion_r1240010016
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/917963a0a39068f2bb3a8df42e257badcfed5e57/zkevm-circuits/src/tx_circuit/sign_verify.rs#L583C1-L592C16

About the other options:
1. The msg_hash 32 byte cells live in the main_gate, so it's easier to do the aggregation via the `main_gate` (option 2).
2. I think that's fine.  Technically there will be (16 + 1) * 2 extra cells used in the main_gate.  That's for each lo/hi, 16 for the byte decomposition copied cells, and 1 for the composed value.  For each signature verification, the ECDSA chip uses like ~100k rows, so I think this extra number rows is negligible :)
3. You could do that, allocating 32 cells in some column of the TxCircuit, but it's just easier to "connect" the TxCircuit columns to the main gate with lo/hi (2 copy constraints, and 2 cells) than 32.
4. You could do that with a copy constraint, (like option 3), or polynomial constraints `lhs - rhs == 0`.  I think copy constraint here is much simpler, because it's the natural API of the `main_gate`.

Reply:
Note that the API of the `main_gate` (and halo2wrong) is very different to the way we have written our circuits in the zkevm project; it can take some time to become familiar with it!

Reply:
fixed in [2629dcf](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1418/commits/2629dcf0d4658ed3cea836db3976c02457bfbf62). Hopefully it was well-addressed this time.

Reply:
One thing I'm not quite sure is, I guess we don't need assign `msg_hash` again like [here](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1418/files#diff-7ba26777eb0846c0f1e6701a9c33cbae5b33294b30b47446904ef66e02133ea8R633). We just need to do like this, right?
```rust
Word::new([msg_hash_cell_lo, msg_hash_cell_hi])
```

Reply:
It's not yet correct :sweat_smile:
We have the ECDSA Chip which checks that a `msg_hash` and a signature (`r, s`) was signed by a key with `pk`.  Then we connect the `msg_hash` and `pk` (which are non native field, so they have their own format) to their corresponding version in `bytes`, and we expose them in `AssignedECDSA`.  So on `AssignedECDSA` we have the signals corresponding to `msg_hash` and `pk`.

Then we connect the `pk` to a keccack check, and connect 20 bytes of the result to the address that is used by the Tx (caller_address).

And also we connect the `msg_hash` to the one used in the Tx.

When I say connect I mean that there are constraints that bind the value of signals (cells) to other signals (cells).

In your latest update you're not using `assigned_ecdsa.msg_hash`, so the signals that contain the `msg_hash` that the ECDSA Chip has used are disconnected from the `msg_hash` that is used in the Tx.

I've fixed that in this commit, please take a look! https://github.com/privacy-scaling-explorations/zkevm-circuits/commit/846b2ef219e70db6c5bbcd1ee86dabce9842a0ee

The main difference is that you were doing `range_chip.decompose` which assigns the `msg_hash` bytes, and builds the `lo, hi`.  Here you're doing a new assignment disconnected from previous cells.  Instead I'm doing `main_gate.decompose` from the already assigned `msg_hash` byte cells that the ECDSA Chip has verified, and from that build `lo, hi`.